metals/src/main/scala/scala/meta/internal/metals/JavaInteractiveSemanticdb.scala

```diff
@@ -0,0 +1,227 @@
+
+class JavaInteractiveSemanticdb(
+    javac: AbsolutePath,
+    jdkVersion: JavaInteractiveSemanticdb.JdkVersion,
+    pluginJars: List[Path],
+    workspace: AbsolutePath,
+    buildTargets: BuildTargets
+) {
+
+  private val readonly = workspace.resolve(Directories.readonly)
+
+  def textDocument(source: AbsolutePath, text: String): s.TextDocument = {
+    val sourceRoot = source.parent
+    val workDir = AbsolutePath(
+      Files.createTempDirectory("metals-javac-semanticdb")
+    )
+    val targetRoot = workDir.resolve("target")
+
+    val targetClasspath = buildTargets
+      .inferBuildTarget(sourceRoot)
+      .flatMap(buildTargets.targetJarClasspath)
+      .getOrElse(Nil)
+      .map(_.toString)
+
+    val jigsawOptions = addExportsFlags ++ patchModuleFlags(source, sourceRoot)
+    val mainOptions =
+      List(
+        javac.toString,
+        "-cp",
+        (pluginJars ++ targetClasspath).mkString(":"),
+        "-d",
+        targetRoot.toString,
+        s"-Xplugin:semanticdb -sourceroot:${sourceRoot} -targetroot:${targetRoot}"
+      )
+    val cmd = mainOptions ::: jigsawOptions ::: source.toString :: Nil
+
+    val stdout = List.newBuilder[String]
+    val ps = SystemProcess.run(
+      cmd,
+      workDir,
+      false,
+      Map.empty,
+      outLine => stdout += outLine,
+      errLine => stdout += errLine
+    )
+
+    val future = ps.complete.recover { case NonFatal(e) =>
+      scribe.error("Running javac-semanticdb failed", e)
+      1
+    }
+
+    val exitCode = Await.result(future, 10.seconds)
+    val semanticdbFile =
+      targetRoot
+        .resolve("META-INF")
+        .resolve("semanticdb")
+        .resolve(s"${source.filename}.semanticdb")
+
+    val doc = if (semanticdbFile.exists) {
+      FileIO
+        .readAllDocuments(semanticdbFile)
+        .headOption
+        .getOrElse(s.TextDocument())
+    } else {
+      val log = stdout.result()
+      if (exitCode != 0 || log.nonEmpty)
+        scribe.warn(s"Javac-semanticdb output:\n${log.mkString("\n")}")
+      s.TextDocument()
+    }
+
+    val out = doc.copy(
+      uri = source.toURI.toString(),
+      text = text,
+      md5 = MD5.compute(text)
+    )
+
+    workDir.deleteRecursively()
+    out
+  }
+
+  private def patchModuleFlags(
+      source: AbsolutePath,
+      sourceRoot: AbsolutePath
+  ): List[String] = {
+    // Jigsaw doesn't allow compiling source with package
+    // that is declared in some existing module.
+    // It fails with: `error: package exists in another module: $packageName`
+    // but it might be fixed by passing `--patch-module $moduleName=$sourceRoot` option.
+    //
+    // Currently there is no infrastucture to detect if package belong to jigsaw module or not
+    // so this case is covered only for JDK sources.
+    if (jdkVersion.hasJigsaw) {
+      source.toRelativeInside(readonly) match {
+        case Some(rel) =>
+          val names = rel.toNIO.iterator().asScala.toList.map(_.filename)
```

Comment:
This might not work with the changes in the Virtual docs PR. We can for sure fix it there afterwards, but maybe it would be possible to prepare for it here?

Reply:
If we aren't going to include Virtual docs to the next release let's keep as it is.
I don't know how to write tests on virtual docs without actual virtual docs impl.